src/matrix/wishart.jl

```diff
@@ -66,27 +77,37 @@ show(io::IO, d::Wishart) = show_multline(io, d, [(:df, d.df), (:S, Matrix(d.S))]
 
 function convert(::Type{Wishart{T}}, d::Wishart) where T<:Real
     P = convert(AbstractArray{T}, d.S)
-    Wishart{T, typeof(P)}(T(d.df), P, T(d.logc0))
+    Wishart{T, typeof(P), typeof(d.rank)}(T(d.df), P, T(d.logc0), d.rank, d.singular)
 end
-function convert(::Type{Wishart{T}}, df, S::AbstractPDMat, logc0) where T<:Real
+function convert(::Type{Wishart{T}}, df, S::AbstractPDMat, logc0, rnk, singular) where T<:Real
     P = convert(AbstractArray{T}, S)
-    Wishart{T, typeof(P)}(T(df), P, T(logc0))
+    Wishart{T, typeof(P), typeof(rnk)}(T(df), P, T(logc0), rnk, singular)
 end
 
 #  -----------------------------------------------------------------------------
 #  Properties
 #  -----------------------------------------------------------------------------
 
-insupport(::Type{Wishart}, X::Matrix) = isposdef(X)
-insupport(d::Wishart, X::Matrix) = size(X) == size(d) && isposdef(X)
+insupport(::Type{Wishart}, X::AbstractMatrix) = ispossemdef(X)
+function insupport(d::Wishart, X::AbstractMatrix)
+    size(X) == size(d) || return false
+    if d.singular
+        return ispossemdef(X, rank(d))
+    else
+        return isposdef(X)
+    end
+end
 
 dim(d::Wishart) = dim(d.S)
 size(d::Wishart) = (p = dim(d); (p, p))
-rank(d::Wishart) = dim(d)
+rank(d::Wishart) = d.rank
 params(d::Wishart) = (d.df, d.S)
 @inline partype(d::Wishart{T}) where {T<:Real} = T
 
-mean(d::Wishart) = d.df * Matrix(d.S)
+function mean(d::Wishart)
```

Comment:
The mean of a singular Wishart is just nV also in the singular case

Reply:
To see this: It is just using the additivity of the covariance of independent centred Gaussians.

Reply:
Shouldn't the expectation over semidefinite matrices be a semidefinite matrix?

Reply:
The sum of only semi-definite matrices can be positive definite. Consider diagonal semi-definite matrices and their sums to see this.

Reply:
The sum of only semi-definite matrices can be positive definite. Consider diagonal semi-definite matrices and their sums to see this.

Reply:
Sure, but you wouldn't report that the mean of a k = 0 von Mises-Fisher distribution is 0, even though that's what sample averages converge to.

Reply:
Hehe, I would :-)

Reply:
Fair enough.